### PR TITLE
Bdog 461 rebase

### PR DESCRIPTION
**Pull request by colin-lamed**
- Ability to register extra codecs for entity leaf filtering and updates. (with the exception of Sum-types)
- Fix parsing of Mongo numbers (there still exists some edge-cases to be analysed).

### CODE REVIEW

File: hmrc-mongo-metrix-play-26/src/main/scala/uk/gov/hmrc/metrix/persistence/MongoMetricRepository.scala [hunk 1]
```diff
@@ -27,8 +27,11 @@ import uk.gov.hmrc.mongo.play.PlayMongoCollection
...
 
 class MongoMetricRepository(collectionName: String = "metrics", mongo: MongoComponent)(implicit ec: ExecutionContext)
-  extends PlayMongoCollection[PersistedMetric](collectionName = collectionName, mongoComponent = mongo, domainFormat = PersistedMetric.format,
-    indexes = Seq(
+  extends PlayMongoCollection[PersistedMetric](
+    collectionName = collectionName,
+    mongoComponent = mongo,
+    domainFormat   = PersistedMetric.format,
+    indexes        = Seq(
       IndexModel(ascending("name"), IndexOptions().name("metric_key_idx").unique(true).background(true))
```
Comment (chotai) [inline hmrc-mongo-metrix-play-26/src/main/scala/uk/gov/hmrc/metrix/persistence/MongoMetricRepository.scala]: Know this isn't one of yours, but is it worth changing this to be camelCased for consistency with the other indexes that we have?

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala [hunk 2]
```diff
@@ -35,18 +38,97 @@ trait Codecs {
...
+        case JsBoolean(b)                     => BsonBoolean.valueOf(b)
+        case JsNumber(n) if n.isValidInt      => new BsonInt32(n.intValue)
+        case JsNumber(n) if n.isValidLong     => new BsonInt64(n.longValue)
+        case JsNumber(n) if n.isDecimalDouble => new BsonDouble(n.doubleValue)
+        case JsNumber(n)                      => //println(s"Converting Number: $n");
+                                                 val res = new BsonDecimal128(new Decimal128(n.bigDecimal))
+                                                 // How to handle `java.lang.NumberFormatException, with message: Conversion to Decimal128 would require inexact rounding of -4.2176255923279509728936555398034786404E-54.`
+                                                 // Alternative format? For now, avoiding in BigDecimal test data generation.
+                                                 //println(s"Converted $res");
+                                                 res
```
Comment (chotai) [inline hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala]: might be worth tidying this up

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala [hunk 3]
```diff
@@ -35,18 +38,97 @@ trait Codecs {
...
+                                                     // should use default conversion to Json. Then PlayJsonReaders will then convert as appropriate
+                                                     // The exception are numbers handled above (otherwise precision of $numberDecimal will be lost)
+                                                     fromJsonDefault(o)
+                                                   else
+                                                     new BsonDocument(
+                                                       o.fields.map { case (k, v) =>
+                                                         new BsonElement(k, jsonToBson(v))
+                                                       }.asJava
+                                                     )
+                                                 //println(s"Converted $res")
```
Comment (chotai) [inline hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala]: same as above

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala [hunk 4]
```diff
@@ -35,18 +38,97 @@ trait Codecs {
...
+        case d : BsonDouble     => JsNumber(d.getValue)
+        case bd: BsonDecimal128 => JsNumber(bd.getValue.bigDecimalValue) // * @throws ArithmeticException if the Decimal128 value is NaN, Infinity, -Infinity, or -0, none of which can be represented as a
+        case s : BsonString     => JsString(s.getValue)
+        case d : BsonDocument   => JsObject(
+                                     d.asScala.map { case (k, v) => (k, bsonToJson(v)) }
+                                   )
+        case other              => // other types, attempt to convert to json object (Strict = `MongoDB Extended JSON format`)
+                                   toJsonDefault(other, JsonMode.STRICT) match {
+                                     case JsDefined(s)   => /*println(s"Converted $other to $s");*/ s
+                                     case _: JsUndefined => println(s"Could not convert $other to Json"); JsNull // TODO logger
```
Comment (chotai) [inline hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/Codecs.scala]: same

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala [hunk 5]
```diff
@@ -34,24 +34,7 @@ trait MongoJavatimeFormats {
 
   val instantFormats: Format[Instant] = Format(instantReads, instantWrites)
 
-  val localDateAsStringWrites: Writes[LocalDate] = new Writes[LocalDate] {
```
Comment (chotai) [inline hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala:37]: nice one! I wanted to remove these too

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala [hunk 6]
```diff
@@ -80,9 +65,9 @@ trait MongoJavatimeFormats {
   val localDateTimeFormats = Format(localDateTimeRead, localDateTimeWrite)
 
   trait Implicits {
-    implicit val instantFormats: Format[Instant]             = outer.instantFormats
-    implicit val localDateFormats: Format[LocalDate]         = outer.localDateFormats
-    implicit val localDateTimeFormats: Format[LocalDateTime] = outer.localDateTimeFormats
+    implicit val jatInstantFormats: Format[Instant]             = outer.instantFormats
```
Comment (chotai) [inline hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala:68]: what's `jat`?

File: hmrc-mongo-test/src/main/scala/uk/gov/hmrc/mongo/test/DefaultMongoCollectionSupport.scala [hunk 7]
```diff
@@ -18,9 +18,12 @@ package uk.gov.hmrc.mongo.test
 
 import com.mongodb.MongoQueryException
 import org.scalatest._
+import scala.concurrent.duration.DurationInt
 
 trait DefaultMongoCollectionSupport extends CleanMongoCollectionSupport with IndexedMongoQueriesSupport {
   this: TestSuite =>
+
+  override implicit val patienceConfig = PatienceConfig(timeout = 30.seconds, interval = 100.millis)
```
Comment (chotai) [inline hmrc-mongo-test/src/main/scala/uk/gov/hmrc/mongo/test/DefaultMongoCollectionSupport.scala:25]: hmm thought this was on master already (with 5 seconds)

**Review by chotai — CHANGES_REQUESTED** [review]
Can you please run `scalafmt`

File: hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala (hunk 6 above)
Comment (colin-lamed) [reply hmrc-mongo-play-26/src/main/scala/uk/gov/hmrc/mongo/play/json/MongoJavatimeFormats.scala:68]: implicits need unique names - the name is unimportant (non-implicit usage should be using the MongoJavatimeFormats.instantFormats).
I've just gone with jot for jodatime and jat for javatime

**Review by chotai — APPROVED** [review]

**Pull request merged**